backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/service/RemittanceAdviceBankingService.java

```diff
@@ -0,0 +1,350 @@
+					if (invoiceProcessingFeeCalculation != null)
+					{
+						final RemittanceAdviceLine remittanceAdviceLine = remittanceAdviceLineMap.get(invoiceProcessingFeeCalculation.getInvoiceId());
+						if (remittanceAdviceLine != null)
+						{
+							final InvoiceId serviceFeeInvoiceId = serviceFeeInvoiceIdsByAssignedInvoiceId.get(invoiceProcessingFeeCalculation.getInvoiceId().getRepoId());
+
+							final RemittanceAdviceLineServiceFee remittanceAdviceLineServiceFee = getRemittanceAdviceLineServiceFee(serviceFeeInvoiceId);
+
+							remittanceAdviceLine.setServiceFeeDetails(remittanceAdviceLineServiceFee);
+						}
+					}
+				});
+	}
+
+	@NonNull
+	private RemittanceAdviceLineServiceFee getRemittanceAdviceLineServiceFee(@NonNull final InvoiceId serviceFeeInvoiceId)
+	{
+		TaxId serviceFeeTaxId = null;
+		ProductId serviceFeeProductId = null;
+		BigDecimal serviceFeeTaxVATRate = null;
+		final List<I_C_InvoiceLine> serviceFeeInvoiceLines = getInvoiceLines(serviceFeeInvoiceId);
+
+		if (!CollectionUtils.isEmpty(serviceFeeInvoiceLines))
+		{
+			final I_C_InvoiceLine firstInvoiceLine = serviceFeeInvoiceLines.get(0);
+			serviceFeeTaxId = TaxId.ofRepoId(firstInvoiceLine.getC_Tax_ID());
+			serviceFeeProductId = ProductId.ofRepoIdOrNull(firstInvoiceLine.getM_Product_ID());
+
+			final Tax serviceFeeTax = taxDAO.getTaxById(serviceFeeTaxId.getRepoId());
+			serviceFeeTaxVATRate = serviceFeeTax.getRate();
+		}
+
+		final I_C_Invoice serviceFeeInvoice = invoiceDAO.getByIdInTrx(serviceFeeInvoiceId);
+
+		return RemittanceAdviceLineServiceFee.builder()
+				.serviceFeeInvoiceId(serviceFeeInvoiceId)
+				.serviceProductId(serviceFeeProductId)
+				.serviceBPartnerId(BPartnerId.ofRepoId(serviceFeeInvoice.getC_BPartner_ID()))
+				.serviceFeeTaxId(serviceFeeTaxId)
+				.serviceVatRate(serviceFeeTaxVATRate)
+				.build();
+	}
+
+	private Map<InvoiceId, RemittanceAdviceLine> getRemittanceAdviceLinesByInvoiceId(final List<RemittanceAdviceLine> remittanceAdviceLines)
+	{
+		return remittanceAdviceLines
+				.stream()
+				.filter(remittanceAdviceLine -> remittanceAdviceLine.getInvoiceId() != null)
+				.collect(Collectors.toMap(RemittanceAdviceLine::getInvoiceId, remittanceAdviceLine -> remittanceAdviceLine));
+	}
+
+	@NonNull
+	private Map<InvoiceId, I_C_Invoice> getInvoiceMapById(@NonNull final List<RemittanceAdviceLine> remittanceAdviceLines)
+	{
+		final ImmutableSet<InvoiceId> invoiceIds = remittanceAdviceLines
+				.stream()
+				.map(RemittanceAdviceLine::getInvoiceId)
+				.filter(Objects::nonNull)
+				.collect(ImmutableSet.toImmutableSet());
+
+		return getInvoicesById(invoiceIds);
+	}
+
+	@NonNull
+	private Map<InvoiceId, I_C_Invoice> getInvoicesById(final Set<InvoiceId> invoiceIds)
+	{
+		return invoiceDAO.getByIdsInTrx(invoiceIds)
+				.stream()
+				.collect(Collectors.toMap(invoice -> InvoiceId.ofRepoId(invoice.getC_Invoice_ID()), invoice -> invoice));
```

Comment:
just a matter of taste, but in this case u could use Maps.uniqueIndex

Reply:
Hi @teosarca FYI so far I just moved the code out of the process-class. That's why it's now popping up into this PR. I'm going to adress as much of those items as possible while getting it to create the proper allocations